### PR TITLE
Adds Notifications Control Center

### LINKED ISSUE #2525: Add Notifications Settings Panel
A new notifications panel is coming. It will add the ability to manage notifications for the notifications stream, email, and app push notifications (even multiple devices) from one place.

### PR DESCRIPTION
**Pull request by roundhill**
Adds an all-new notifications settings area that lets you finely tune the notifications settings, per blog.

Settings are toggled in the `NotificationsSettingsDialogPreference` dialog. When that dialog is closed, the changed settings are posted back to WP.com to be saved. Settings should match up with what you see on `/me/notifications` on wpcalypso.

Also adds a new `Notification Setting` row to `MeFragment` and removes it from the Account Settings fragment. There's also a new check to see if the global app notifications settings has been disabled. If so, certain rows in the notifications settings should be disabled.

Note: The new settings are not being enforced on WP.com yet, so you won't be able to test that you actually stop receiving notifications that you disable. But, the settings should save and persist.

Fixes #2525

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsFragment.java [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import javax.annotation.Nonnull;
+
+import de.greenrobot.event.EventBus;
+
+public class NotificationsSettingsFragment extends PreferenceFragment {
+    public static final String TAG = "NotificationSettingsFragment";
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsFragment.java]: `TAG` is unused.

File: WordPress/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -599,7 +595,65 @@
...
+    <string name="notifications_account_emails">Updates from WordPress.com</string>
+    <string name="notifications_wpcom_updates">WordPress.com Updates</string>
+    <string name="notifications_other_blogs">Other Sites</string>
+    <string name="notifications_comments_other_blogs">Comments on other sites</string>
+    <string name="notifications_tab">Notifications tab</string>
+    <string name="email">Email</string>
+    <string name="app_notifications">App notifications</string>
+    <string name="comment_likes">Comment likes</string>
+    <string name="replies_to_your_comments">Replies to your comments</string>
+    <string name="error_loading_notifications">Could not load notifications settings</string>
```
Comment (nbradbury) [inline WordPress/src/main/res/values/strings.xml]: "Notifications" should be singular to match the activity title ("Notification Settings").

**Comment by nbradbury** [comment]
Could uses some extra whitespace after the dialog title.

device-2015-08-05-175218

File: WordPress/src/main/java/org/wordpress/android/WordPressDB.java [hunk 3]
```diff
@@ -430,13 +430,25 @@ public boolean addBlog(Blog blog) {
     }
 
     public List<Map<String, Object>> getBlogsBy(String byString, String[] extraFields) {
-        return getBlogsBy(byString, extraFields, 0);
+        return getBlogsBy(byString, extraFields, 0, true);
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/WordPressDB.java]: Are you sure it's safe to assume every place that calls this wants to hide Jetpack blogs?

**Comment by nbradbury** [comment]
After I added my self-hosted blog to the app, it showed up twice in notification settings.

device-2015-08-05-192256

File: WordPress/src/main/java/org/wordpress/android/WordPressDB.java [hunk 4]
```diff
@@ -430,13 +430,25 @@ public boolean addBlog(Blog blog) {
...
+    public List<Map<String, Object>> getBlogsBy(String byString, String[] extraFields,
+                                                int limit, boolean shouldHideJetpack) {
         if (db == null) {
-            return new Vector<Map<String, Object>>();
+            return new Vector<>();
         }
+
+        if (shouldHideJetpack) {
+            // Hide Jetpack blogs that were added in FetchBlogListWPCom
+            String hideJetpackArgs = "NOT(dotcomFlag=0 AND wpVersion='' AND blogId!=1)";
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/WordPressDB.java]: Could use a comment here explaining why `blogId!=1` is necessary.

File: WordPress/src/main/java/org/wordpress/android/WordPressDB.java (hunk 3 above)
Comment (roundhill) [inline WordPress/src/main/java/org/wordpress/android/WordPressDB.java]: Yeah, no other areas of the app would want to access the Jetpack blogs that were added using `FetchBlogListWPCom`. The auth token won't work for those sites so they should remain hidden everywhere except for Notification Settings.

**Comment by nbradbury** [comment]
I'm seeing an empty "WordPress.com Updates" section now.

device-2015-08-06-183622

**Comment by nbradbury** [comment]
If I change a notification setting but cancel out of the dialog, the setting still sticks. [Demo video](https://cloudup.com/c6IYp7pUQc8).

File: WordPress/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -599,7 +595,68 @@
...
     <string name="error_notification_open">Could not open notification</string>
     <string name="ignore">Ignore</string>
-    <string name="push_auth_expired">The request has expired. Sign in to WordPress.com to try again.</string>
+    <string name="push_auth_expired">The request has expired. Sign in to WordPress.com to try again.</string><string name="notifications_updates_from_wordpress">Updates from WordPress.com</string>
+
+    <!-- Notification Settings -->
+    <string name="notification_settings">Notification Settings</string>
+    <string name="notifications_sights_and_sounds">Sights and Sounds</string>
+    <string name="your_sites">Your Sites</string>
+    <string name="notifications_account_emails_summary">We’ll always send important emails regarding your account, security, privacy, and purchase transactions, but you can get some fun extras, too!</string>
```
Comment (nbradbury) [inline WordPress/src/main/res/values/strings.xml]: I know I'm harping about wording, but this "fun extras" part sounds like insincere marketing-speak. Any chance we can drop that?

Comment (roundhill) [inline WordPress/src/main/res/values/strings.xml]: I think the marketing team wrote that ;)

**Comment by roundhill** [comment]
> If I change a notification setting but cancel out of the dialog, the setting still sticks. Demo video.

Indeed! Fixed in https://github.com/wordpress-mobile/WordPress-Android/commit/b7079e36c1c143fc4f69e41c84e056555088377d

**Comment by roundhill** [comment]
> I'm seeing an empty "WordPress.com Updates" section now.

Do you see this everytime? I haven't reproduced it yet, but I'm guessing it's a race condition.

**Comment by nbradbury** [comment]
> Do you see this everytime?

I can't repro it now, and I tried several times.

**Comment by nbradbury** [comment]
device-2015-08-07-121812

I think this section violates a couple of things from [Google's guidelines about settings](https://www.google.com/design/spec/patterns/settings.html#settings-labels-secondary-text):

* Avoid repeating words from the label in secondary text
* If the label is sufficient on its own, don't add a secondary text description

Is it even necessary to have these as two groups? Seems like a single "Other" group would suffice, without any secondary text, ie:

updated

**Comment by roundhill** [comment]
> Is it even necessary to have these as two groups? Seems like a single "Other" group would suffice, without any secondary text, ie:

Looks ok to me, what say you @drw158?

**Comment by drw158** [comment]
The repeating words is a good point. After taking a look at iOS as well, I didn't realize how much we were repeating. I think an "Other" heading would be an improvement. "Comments on other sites" probably doesn't need any more explanation, but Updates does. FWIW, I just noticed this but I shortened the text for the native apps to be:

>We’ll always send important emails regarding your account, but you can get some fun extras, too!

**Comment by roundhill** [comment]
> I think an "Other" heading would be an improvement.

OK, thanks @drw158. Here's how that looks now:
screen shot 2015-08-10 at 3 52 59 pm

**Comment by nbradbury** [comment]
FWIW, I think that's a whole lot simpler. But it sure would be nice if we could drop that cheesy "fun extras" wording at the same time. It would be hard for me to overstate how much I dislike that :)

**Comment by drw158** [comment]
We got the green light to change the copy to:

> We’ll always send important emails regarding your account, but you can get some helpful extras, too.

**Comment by nbradbury** [comment]
The wording is much better now, thanks for putting up with my complaints :) Overall this looks good, but the first time I tried the latest changes the "Updates from WordPress.com" setting was missing again. I can't repro it, but there's obviously a bug here somewhere.

device-2015-08-11-181437

**Comment by roundhill** [comment]
> there's obviously a bug here somewhere

I wasn't able to reproduce, but give it another pull and go. I tweaked how some of the JSON objects are retrieved and removed some unnecessary `null` checks.

**Comment by nbradbury** [comment]
> give it another pull and go

I'm not seeing the bug after pulling these changes, but looking at the code I'm not sure what actually fixed it! Given this, I think it'd be a good idea to add some comments to `NotificationsSettings.java` to help the dev who eventually inherits this code :)

**Comment by roundhill** [comment]
> I think it'd be a good idea to add some comments to NotificationsSettings.java

Added some (hopefully helpful) comments in https://github.com/wordpress-mobile/WordPress-Android/commit/98495ef4912972b054ca2a2d9e2a81323bd8cee9

**Comment by nbradbury** [comment]
:shipit:

**Pull request merged**